### PR TITLE
Add rmw_publisher_wait_for_all_acked

### PR DESCRIPTION
**Pull request by Barry-Xu-2018**
Related to ros2/rmw#295

### CODE REVIEW

**Review by ahcorde — APPROVED** [review]

**Review by fujitatomoya — APPROVED** [review]
lgtm

**Review by ivanpauno — APPROVED** [review]

**Comment by ivanpauno** [comment]
Could you add a test case in https://github.com/ros2/rmw_implementation/tree/master/test_rmw_implementation?

I don't need a test that actually confirms that all messages were acked (which is a bit hard to test), only test basic things:
e.g. passing a nullptr publisher returns an error, passing a valid BEST_EFFORT publisher always return immediatly RMW_RET_OK, etc.

**Comment by Barry-Xu-2018** [comment]
> Could you add a test case in https://github.com/ros2/rmw_implementation/tree/master/test_rmw_implementation?
> 
> I don't need a test that actually confirms that all messages were acked (which is a bit hard to test), only test basic things:
> e.g. passing a nullptr publisher returns an error, passing a valid BEST_EFFORT publisher always return immediatly RMW_RET_OK, etc.

Okay. I add a simple test case.  
FYI, I add more test cases (such as timeout) in rcl.

File: test_rmw_implementation/test/test_publisher.cpp [hunk 1]
```diff
@@ -527,6 +527,38 @@ TEST_F(CLASSNAME(TestPublisherUse, RMW_IMPLEMENTATION), publish_serialized_with_
...
+    EXPECT_EQ(RMW_RET_OK, ret) << rmw_get_error_string().str;
+  });
+
+  size_t subscription_count = 0u;
+  rmw_ret_t ret = rmw_publisher_count_matched_subscriptions(pub, &subscription_count);
+  ASSERT_EQ(subscription_count, static_cast<size_t>(1));
+
+  ret = rmw_publish(pub, &input_message, nullptr);
+  ASSERT_EQ(ret, RMW_RET_OK);
+
```
Comment (ivanpauno) [inline test_rmw_implementation/test/test_publisher.cpp:553]: This isn't adding much to the test really, I would keep it simple:

```suggestion
```

Comment (Barry-Xu-2018) [reply test_rmw_implementation/test/test_publisher.cpp]: @ivanpauno Addressed your comments. https://github.com/ros2/rmw_implementation/pull/188/commits/87b997a7a46799a786fef5b1f0e81b491b58a70a

**Review by ivanpauno — APPROVED** [review]

**Comment by ivanpauno** [comment]
@ros-pull-request-builder retest this please (this should pass now)

**Comment by ivanpauno** [comment]
@Barry-Xu-2018 can you check the test failures?

**Comment by Barry-Xu-2018** [comment]
@ivanpauno 

The failure is related to `test_rmw_implementation.TestPublisherUse__rmw_cyclonedds_cpp.wait_for_all_acked_with_best_effort `.    
It is because ros2/rmw_cyclonedds#294 isn't merged now.    

Could you check and merge it ?

**Comment by ivanpauno** [comment]
Thanks @Barry-Xu-2018, I was sure I merged that one :smile:

**Comment by Barry-Xu-2018** [comment]
@ivanpauno 

After ros2/rmw_cyclonedds#294 is merged, we should re-execute CI.    
Could you help to execute CI again ?

**Comment by fujitatomoya** [comment]
@ros-pull-request-builder retest this please

CC: @Barry-Xu-2018 @ivanpauno

**Comment by ivanpauno** [comment]
The PR checker should pass after https://github.com/ros/rosdistro/pull/29924 is merged.
No need to run CI again, this is only wrapping new API and adding tests for it.

Thanks for the patience, but merging changes in rmw_* without breaking the rolling builds is a long process ...

**Comment by ivanpauno** [comment]
@ros-pull-request-builder retest this please

**Comment by ivanpauno** [comment]
@ros-pull-request-builder retest this please

**Comment by ivanpauno** [comment]
finally ... :smiley: 
going in!!

**Pull request merged**